src/docs/perf/rendering/shader.md

```diff
@@ -184,17 +184,17 @@ difference as illustrated in the beginning of this article.
    rasterization time in release mode; the worst frame rasterization time is
    a good indicator on how severe the shader compilation jank is.)
 
-3. **SkSL warm-up doesn't help newer iPhones using Metal.**<br><br>
-   Flutter recently migrated from OpenGL to Metal for all newer iOS devices.
-   (Please reference [Metal on iOS FAQ][] on which iOS devices are considered
-   new enough to use Metal.) However, Skia currently only implemented the SkSL
-   warm-up for OpenGL. So the SkSL warm-up would only speed up older iOS devices
-   by default. If you find shader compilation jank to be an issue for your app
-   on newer iPhones, please let us know by filing a [GitHub issue][]. In the
-   longer term, we have a plan to use test-based shader warm-up to mitigate
-   this. If there's an urgent need for fixing shader compilation jank on newer
-   iPhones, please leave feedback on [Issue 61045][], and we can help you turn
-   on OpenGL for your app.
+3. **SkSL warm-up doesn't help on iOS.**<br><br>
+   Flutter uses Metal on most iOS devices. (Please reference [Metal on iOS FAQ][]
+   for a list of which iOS devices use Metal.) Unfortunately, shader warm-up
+   as currently available has proved insufficient to reduce shader compilation
+   jank. This is because shaders go through two compilation steps: SkSL (Skia
```

Comment:
That the shaders go through the two expensive steps doesn't quite answer why it can't be cached and reused. I'd say something like:
```
Metal requires more information to be explicitly specified upfront for the creation of pipelines. This information is only available today in the frame workload which make warmup from SKSL less effective.
```

Reply:
I don't really understand how this explains why shader warm-up can't work on Metal.

Fundamentally, shader warm-up on Metal does work, in the literal sense that if I run the same animation twice, the first time it's slow and is compiling shaders, and the second it's fast and isn't compiling shaders.

So what I don't understand is, why can't we do what we do on the other platforms and compile those shaders we know the application will need (because the developer told us by running the app and finding out what shaders the application will need, as described on this very web page) ahead of time? As if we'd run the animation and compiled those shaders?

Reply:
As per https://github.com/flutter/flutter/issues/32170#issuecomment-790939469 it may in fact be possible to do shader warm-up on metal as well as OpenGL. I'll update this PR once we have more information.